src/pages/index.js

```diff
@@ -83,30 +166,55 @@ function createCard(data) {
 
 function handleAddCardFormSubmit(values) {
   const name = values.title;
+  // console.log(values);
   const link = values.URL;
   // generateCard({ name, link }, cardsWrap);
-  const cardEl = createCard({ name, link });
-  cardSection.addItem(cardEl);
+  api.addCards({ title: values.title, url: values.URL }).then((res) => {
+    newCardPopup.setButtonText(true, "Saving...");
+
+    // res is the card data
+    const cardEl = createCard(res);
+    cardSection.addItem(cardEl);
 
-  newCardPopup.close();
+    newCardPopup.close();
+  });
 }
 
 function handleProfileFormSubmit(values) {
-  userInfo.setUserInfo(values);
-  editProfilePopup.close();
+  api.updateUserInfo(values).then((res) => {
+    editProfilePopup.setButtonText(true, "Saving...");
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```